### ISSUE TITLE
Possibly incompatible with New Relic agent (or the other way around.)

### ISSUE THREAD

**Issue by nathanpalmer** [saturnflyer/casting#114]
We recently integrated New Relic into our application to test out some metrics and it initially connects but starts throwing errors to which Casting is a big part of the stack trace.

I wanted to get some thoughts on this to find out if anyone else is using New Relic and whether or not this has been an issue.

*RuntimeError: can't modify frozen NewRelic::Agent::Configuration::ServerSource*
```
/var/bundle/ruby/2.2.0/gems/casting-0.7.1/lib/casting/missing_method_client.rb:50:in `__delegates__'
/var/bundle/ruby/2.2.0/gems/casting-0.7.1/lib/casting/missing_method_client.rb:67:in `method_delegate'
/var/bundle/ruby/2.2.0/gems/casting-0.7.1/lib/casting/missing_method_client.rb:63:in `respond_to_missing?'
/var/bundle/ruby/2.2.0/gems/newrelic_rpm-3.16.0.318/lib/new_relic/agent/configuration/manager.rb:99:in `respond_to?'
/var/bundle/ruby/2.2.0/gems/newrelic_rpm-3.16.0.318/lib/new_relic/agent/configuration/manager.rb:99:in `block in source'
/var/bundle/ruby/2.2.0/gems/newrelic_rpm-3.16.0.318/lib/new_relic/agent/configuration/manager.rb:98:in `each'
/var/bundle/ruby/2.2.0/gems/newrelic_rpm-3.16.0.318/lib/new_relic/agent/configuration/manager.rb:98:in `source'
/var/bundle/ruby/2.2.0/gems/newrelic_rpm-3.16.0.318/lib/new_relic/agent/transaction_sample_builder.rb:120:in `transaction_trace_threshold'
/var/bundle/ruby/2.2.0/gems/newrelic_rpm-3.16.0.318/lib/new_relic/agent/transaction_sample_builder.rb:111:in `finish_trace'
/var/bundle/ruby/2.2.0/gems/newrelic_rpm-3.16.0.318/lib/new_relic/agent/transaction_sampler.rb:104:in `on_finishing_transaction'
/var/bundle/ruby/2.2.0/gems/newrelic_rpm-3.16.0.318/lib/new_relic/agent/transaction.rb:519:in `commit!'
/var/bundle/ruby/2.2.0/gems/newrelic_rpm-3.16.0.318/lib/new_relic/agent/transaction.rb:503:in `stop'
/var/bundle/ruby/2.2.0/gems/newrelic_rpm-3.16.0.318/lib/new_relic/agent/transaction.rb:148:in `stop'
/var/bundle/ruby/2.2.0/gems/newrelic_rpm-3.16.0.318/lib/new_relic/agent/instrumentation/middleware_tracing.rb:108:in `call'
/var/bundle/ruby/2.2.0/gems/puma-2.15.3/lib/puma/configuration.rb:79:in `call'
/var/bundle/ruby/2.2.0/gems/puma-2.15.3/lib/puma/server.rb:541:in `handle_request'
/var/bundle/ruby/2.2.0/gems/puma-2.15.3/lib/puma/server.rb:388:in `process_client'
/var/bundle/ruby/2.2.0/gems/puma-2.15.3/lib/puma/server.rb:270:in `block in run'
/var/bundle/ruby/2.2.0/gems/puma-2.15.3/lib/puma/thread_pool.rb:106:in `call'
/var/bundle/ruby/2.2.0/gems/puma-2.15.3/lib/puma/thread_pool.rb:106:in `block in spawn_thread'
```

The lines in new_relic come down to this

```ruby
def source(key)
  config_stack.each do |config|
    if config.respond_to?(key.to_sym) || config.has_key?(key.to_sym)
      return config
    end
  end
end
```
 and on the `config.respond_to?` method which seems to get into the fact that it's trying to initialize @__delegates__ on a frozen object.

**Comment by nathanpalmer** [comment]
Digging in further this seems to come down to the fact that we've included Casting::Client in the base Hash object (we mixin things to hashes for added functionality) and of course New Relic will freeze these objects.

This bit of code fails on the respond_to? line just like new relic.

https://ide.c9.io/nathanpalmer/casting-newrelic

```ruby
require 'casting'

puts "Creating Class"
class Something < ::Hash
  include Casting::Client
  delegate_missing_methods
end

puts "Creating object"
something = Something.new

puts "Freezing object"
something.freeze

puts "Checking respond_to?"
something.respond_to?(:thing)

puts "Done"
```

**Comment by nathanpalmer** [comment]
FYI, this can be fixed by checking whether or not the object is frozen before attempting to create the `@__delegates__` property.

```ruby
module Casting
  module MissingMethodClient
    private
      def __delegates__
        if self.frozen? && @__delegates__.nil?
          return []
        end
        
        super
      end
  end
end
```

**Comment by saturnflyer** [comment]
thanks for finding this @nathanpalmer.
I committed a test case here https://github.com/saturnflyer/casting/commit/953bf4eb1fca3bb8e794437197449b1a32628014
Leaving this open for now since I'm not sure if that test is wide enough. I'd love suggestions if you have anything further.
I'll release a new version of the gem soon

**Comment by saturnflyer** [comment]
@nathanpalmer released 0.7.2 with these changes

**Comment by nathanpalmer** [comment]
Thanks @saturnflyer!
